deps/chakrashim/src/v8handlescope.cc

```diff
@@ -54,28 +54,43 @@ HandleScope *HandleScope::GetCurrent() {
 }
 
 bool HandleScope::AddLocal(JsValueRef value) {
-  if (_count < _countof(_locals)) {
+  if (_count < kOnStackLocals) {
     _locals[_count++] = value;
     return true;
   }
 
+  // Ran out of _locals. Push everything in array and empty
+  // the _locals array.
   if (_refs == JS_INVALID_REFERENCE) {
     if (JsCreateArray(1, &_refs) != JsNoError) {
       return AddLocalAddRef(value);
     }
   }
 
-  JsValueRef index;
+  JsValueRef saveInHandleScopeFunction =
+    jsrt::ContextShim::GetCurrent()->GetsaveInHandleScopeFunction();
 
-  if (JsIntToNumber(_count - _countof(_locals), &index) != JsNoError) {
-    return AddLocalAddRef(value);
+  // kOnStackLocals + 1 (undefined) + 1 (javascriptarray) + 1 (current value)
+  JsValueRef args[kOnStackLocals + 3];
+  args[0] = jsrt::GetUndefined();
+  args[1] = _refs;
+  args[2] = value;
+  for (int index = 0; index < _count;index++) {
+    args[index + 3] = _locals[index];
```

Comment:
Thanks for the update! There may be one more small optimization we can do.

I think you can change the shim function to:

```JavaScript
function (...values) {
    this.push(...values);
}
```

Then omit the first `undefined` and use `_refs` array as `this` instead. JsCallFunction(...[_refs, arg0, arg1, ...]).

Next can omit this local array copy. Change cache size `_locals[kOnStackLocals + 1]`. Reserve and use _locals[0] for the _refs array, _locals[1..] for the values (remove _refs field). Now _locals can be directly used to call JsCallFunction.

Change the sequence in this function:
```
If _locals full
  if flush fail
    return AddLocalAddRef(value)
  clear _locals; count = 1

_locals[count++] = value;
return true;
```